course_inventory/published_date.py

```diff
@@ -75,40 +106,113 @@ def published_date_resp_parsing(self, response):
+                responses = [future_session.get(retry_list[course]['url'], headers=headers)
+                             for course in retry_list]
             else:
                 logger.info("Initial Round of Fetching course published date")
                 responses = [
-                    session.get(f'{self.canvas_url}/api/v1/audit/course/courses/{course_id}?per_page=100',
-                                headers=headers)
+                    future_session.get(f'{self.canvas_url}/api/v1/audit/course/courses/{course_id}?per_page=100',
+                                       headers=headers)
                     for course_id in course_ids
                 ]
 
             for response in as_completed(responses):
                 self.published_date_resp_parsing(response)
 
-        if len(self.published_course_next_page_list) != 0:
-            logger.info(f"""Pagination size {len(self.published_course_next_page_list)} with published_at date items
-                         {self.published_course_next_page_list}""")
-            self.get_published_course_date(course_ids, self.published_course_next_page_list)
-
-        return self.published_course_date
+        if len(self.published_date_retry_bucket) != 0:
+            logger.info(f"""Retrying now with list size {len(self.published_date_retry_bucket)} 
+                        {self.published_date_retry_bucket}""")
+            self.get_published_course_date(course_ids, self.published_date_retry_bucket)
+
+    def get_published_date(self) -> pd.DataFrame:
+        pd.set_option('display.max_column', None)
+        logger.info("Getting into fetching published date routine")
+        courses_with_pub_date_col_df = self.filter_courses_to_fetch_published_date()
+        is_published_date_all_empty = courses_with_pub_date_col_df['published_at'].isnull().all()
+        published_date_in_db = courses_with_pub_date_col_df[
+            (courses_with_pub_date_col_df['published_at'].notnull())].shape
+        course_avail_with_no_pub_date_df = courses_with_pub_date_col_df.loc[
+            (courses_with_pub_date_col_df['workflow_state'] == 'available') &
+            (courses_with_pub_date_col_df['published_at'].isnull())].copy(deep=True)
+        course_avail_with_no_pub_date_list = course_avail_with_no_pub_date_df['canvas_id'].to_list()
+        logger.info(f"Published dates going to be fetched are: {len(course_avail_with_no_pub_date_list)}")
+        if len(course_avail_with_no_pub_date_list) == 0:
+            courses_with_pub_date_col_df = courses_with_pub_date_col_df[['canvas_id', 'published_at']]
+            logger.info("No more published date to fetch than what is stored in DB")
+            logger.info(f"Database should have {published_date_in_db[0]} published dates")
+            return courses_with_pub_date_col_df
+
+        self.get_published_course_date(course_avail_with_no_pub_date_list)
+
+        if len(self.published_course_date) > 0:
```

Comment:
Maybe we could add an `else` saying that no new data was collected? I like to start my `if/else` blocks with the fail case so that a short `else` isn't hidden after a bunch of code. So something like this:

```
if len(self.published_course_date) == 0:
   logger.info('No new published dates for courses were collected')
else:
   ... (the rest of your code)
```

Reply:
You are logging the size of `published_course_date` later, so maybe it's not necessary?

Reply:
yeah... exactly